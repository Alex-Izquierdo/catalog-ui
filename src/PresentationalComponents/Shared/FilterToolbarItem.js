import React, { Component } from 'react';
import propTypes from 'prop-types';
import { Section } from '@red-hat-insights/insights-frontend-components';
import { ToolbarGroup, ToolbarItem, TolbarSection, Title, Button, TextInput } from '@patternfly/react-core';
import { css } from '@patternfly/react-styles';

class FilterToolbarItem extends Component {
    state = {
        searchValue: ''
    };

    render() {
        const { searchValue } = this.state;
        return (
            <ToolbarGroup className="searchToolbar">
                <ToolbarItem>
                    <div className="pf-c-input-group">
                        <TextInput placeholder={ this.props.placeholder } value={ searchValue } type="text" onChange={ this.props.onFilterChange }
                            aria-label="Find product button"></TextInput>
                        <Button variant="tertiary" id="searchProductButton" onClick={ this.onSearchSubmit }>
                            <i className="fas fa-search" aria-hidden="true"></i>
                        </Button>
                    </div>
                </ToolbarItem>
            </ToolbarGroup>);
    };
};

FilterToolbarItem.propTypes = {
    history: propTypes.object,
    onFilterChange: propTypes.func,
    placeholder: propTypes.string
};

export default FilterToolbarItem;

