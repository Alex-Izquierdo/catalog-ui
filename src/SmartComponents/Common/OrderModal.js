import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Modal } from '@patternfly/react-core';

import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../../presentational-components/shared/image-with-default';
import OrderServiceFormStepConfiguration from '../Order/OrderServiceFormStepConfiguration';

const OrderModal = ({ serviceData, closeUrl, history: { push }}) => serviceData ? (
  <Modal
    isOpen
    hideTitle
    onClose={ () => push(closeUrl) }
    style={ { maxWidth: 800, minHeight: 300 } }
  >
    <ImageWithDefault src = { serviceData.imageUrl || CatItemSvg } width="40" />
    { serviceData.name }
    <OrderServiceFormStepConfiguration closeUrl={ closeUrl } { ...serviceData } />
  </Modal>
) : null;

OrderModal.propTypes = {
  orderData: PropTypes.func,
  serviceData: PropTypes.object,
  closeUrl: PropTypes.string.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

const mapStateToProps = ({
  orderReducer: { selectedItem, servicePlans }
}) => ({
  selectedItem,
  servicePlans
});

export default withRouter(connect(mapStateToProps)(OrderModal));
