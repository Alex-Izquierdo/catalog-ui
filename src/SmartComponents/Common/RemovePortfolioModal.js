import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import propTypes from 'prop-types';
import { Stack, StackItem, Button} from '@patternfly/react-core';
import { Main, PageHeader, PageHeaderTitle } from '@red-hat-insights/insights-frontend-components';
import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { removePortfolio, fetchPortfolios } from '../../redux/Actions/PortfolioActions';

class RemovePortfolioModal extends Component {
  onSubmit = () => {
    this.props.removePortfolio(this.props.portfolio.id);
    this.props.closeModal();
  }

  onCancel = () => {
    this.props.addNotification({
      variant: 'warning',
      title: 'Removing portfolio',
      description: 'Removing portfolio was cancelled by the user.'
    });
    this.props.closeModal();
  }

  // Use pf4 components
  render() {
    return (
      <Main title={ 'Remove Portfolio' }>
        <Stack>
          <StackItem>
            <PageHeader>
              <PageHeaderTitle title= 'Remove Portfolio' />
            </PageHeader>
          </StackItem>
          <StackItem>
            <PageHeaderTitle>
              Remove Portfolio
            </PageHeaderTitle>
            <Button variant="primary" type="button" onClick={ this.onSubmit }>
              Remove
            </Button>
            <Button variant="primary" type="button" onClick={ this.onCancel }>
              Cancel
            </Button>
          </StackItem>
        </Stack>
      </Main>
    );
  }
}

const mapStateToProps = ({ portfolioReducer: { isLoading }}) => ({ isLoading });

const mapDispatchToProps = dispatch => bindActionCreators({
  addNotification,
  removePortfolio,
  fetchPortfolios
}, dispatch);

RemovePortfolioModal.propTypes = {
  isLoading: propTypes.bool,
  addNotification: propTypes.func,
  removePortfolio: propTypes.func,
  fetchPortfolios: propTypes.func,
  closeModal: propTypes.func,
  portfolio: propTypes.object
};

export default connect(mapStateToProps, mapDispatchToProps)(RemovePortfolioModal);
