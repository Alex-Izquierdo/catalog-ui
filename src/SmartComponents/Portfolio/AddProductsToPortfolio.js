import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import propTypes from 'prop-types';
import { Section } from '@red-hat-insights/insights-frontend-components';
import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { fetchPlatformItems } from '../../redux/Actions/PlatformActions';
import ContentGallery from '../../SmartComponents/ContentGallery/ContentGallery';
import MainModal from '../Common/MainModal';
import '../Platform/platformitems.scss';
import '../../SmartComponents/Portfolio/portfolio.scss';
import PortfolioOrderToolbar from '../../PresentationalComponents/Portfolio/PortfolioOrderToolbar';
import AddProductsTitleToolbar from '../../PresentationalComponents/Portfolio/AddProductsTitleToolbar';
import PlatformDashboard from '../../PresentationalComponents/Platform/PlatformDashboard';
import PlatformSelectToolbar from '../../SmartComponents/Common/PlatformSelectToolbar';
import { addToPortfolio } from '../../redux/Actions/PortfolioActions';
import PlatformItem from '../../PresentationalComponents/Platform/PlatformItem';

class AddProductsToPortfolio extends Component {
    state = {
      selectedPlatforms: [],
      checkedItems: []
    };

    async fetchData(platforms) {
      await this.props.fetchPlatformItems(platforms);
    }

    onPlatformSelectionChange = (selectedValues) => {
      this.setState({ selectedPlatforms: selectedValues });
      this.fetchData({ platform: selectedValues[0].id });
    };

    onToggleItemSelect = (event) => {
      const item = event.target.id;
      const isChecked = event.target.checked;
      console.log('item select: ', event);
      if (isChecked) {
        this.setState(state => {
          const checkedItems = [ ...state.checkedItems, item ];
          return {
            checkedItems
          };
        });
      }
      else {
        this.setState(prevState => ({ checkedItems: prevState.checkedItems.filter((value) => { return !(value === item.id);}) }));
      }
    };

    onAddToPortfolio = () =>
      this.props.addToPortfolio(this.props.portfolio.id, this.state.checkedItems)
      .then(() => this.props.resetViewMode(null, true));

    render() {
      let filteredItems = [];
      if (this.props.platformItems) {
        filteredItems = {
          items: this.props.platformItems.map(item => (
            <PlatformItem
              key={ item.id }
              { ...item }
              editMode
              onToggleItemSelect={ this.onToggleItemSelect }
              checkedItems={ this.state.checkedItems }
            />)),
          isLoading: this.props.isLoading
        };
      }

      let title = (this.props.portfolio) ? this.props.portfolio.name : '';
      return (
        <Section>
          <PortfolioOrderToolbar/>
          <AddProductsTitleToolbar title={ title }
            onClickAddToPortfolio = { this.onAddToPortfolio }
            onClickCancelAddProducts={ this.props.resetViewMode }/>
          <PlatformSelectToolbar onOptionSelect={ this.onPlatformSelectionChange } { ...this.props } />
          { (this.state.selectedPlatforms.length > 0) &&
                    this.state.selectedPlatforms.map((platform)=> {return (<ContentGallery key={ platform.id } { ...filteredItems }
                      title={ platform.name }
                      editMode = { true }
                      onToggleSelect = { this.onToggleItemSelect }
                      checkedItems = { this.state.checkedItems }
                    />);}) }
          { (this.state.selectedPlatforms.length < 1) && <PlatformDashboard/> }
          <MainModal/>
        </Section>
      );
    }
}

const mapStateToProps = ({ platformReducer: { platformItems, isPlatformDataLoading }}) => ({
  platformItems,
  isLoading: isPlatformDataLoading
});

const mapDispatchToProps = dispatch => bindActionCreators({
  addNotification,
  fetchPlatformItems,
  addToPortfolio
}, dispatch);

AddProductsToPortfolio.propTypes = {
  platformItems: propTypes.array,
  isLoading: propTypes.bool,
  isEditMode: propTypes.bool,
  addToPortfolio: propTypes.func,
  resetViewMode: propTypes.func,
  fetchPlatformItems: propTypes.func,
  portfolio: propTypes.shape({
    name: propTypes.string,
    id: propTypes.oneOfType([ propTypes.string, propTypes.number ]).isRequired
  }).isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(AddProductsToPortfolio);
