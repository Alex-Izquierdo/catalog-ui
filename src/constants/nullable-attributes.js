export const PORTFOLIO_ITEM_NULLABLE = [
  'display_name', 'description', 'long_description', 'distributor', 'documentation_url', 'support_url', 'workflow_ref'
];
