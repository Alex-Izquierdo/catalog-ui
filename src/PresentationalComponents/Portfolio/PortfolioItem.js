import React from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import './portfolioitem.scss';
import propTypes from 'prop-types';
import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../ImageWithDefault';
import { PortfolioStore } from "../../Store/Actions/PortfolioActions";
import { hideModal, showModal } from "../../Store/Actions/MainModalActions";
import { GridItem, Card, CardHeader, CardBody, CardFooter } from '@patternfly/react-core';
import {bindMethods} from "../../Helpers/Shared/Helper";


const propLine = (prop, value) => {
    return(<div className = "card_element"> {value} </div>);
};

const toDisplayProperty = property => {
    return ['description'].includes(property)
};

const propDetails = item => {
    let details = [];

    for (let property in item) {
        if (item.hasOwnProperty(property) && toDisplayProperty(property)) {
            if (item[property] && item[property] !== undefined) {
                details.push(propLine(property, item[property].toString()));
            }
        }
    }
    return details;
};

const itemDetails = props => {
    let details = propDetails(props);
    return (
        <React.Fragment>
            <div>{details}</div>
        </React.Fragment>
    );
};

const mapDispatchToProps = dispatch => {
    return {
        hideModal: () => dispatch(hideModal()),
        showModal: (modalProps, modalType) => {
            dispatch(showModal({ modalProps, modalType }))
        }
    };
};


class PortfolioItem extends React.Component {
    constructor(props) {
        super(props);
        this.handleOnClick = this.handleOnClick.bind(this);
    };

    handleOnClick() {
        console.log( 'Before OrderService');
        this.setState({ showOrder: true });
        this.props.showModal({
            open: true,
            servicedata: this.props,
            closeModal: this.props.hideModal
        }, 'order');
    };

    render() {
        return (
            <GridItem GridItem sm={6} md={4} lg={4} xl={3}>
                <Card>
                    <div onClick={ () => {this.handleOnClick(this.props)}}>
                        <CardHeader className="card_header">
                            <ImageWithDefault src={this.props.imageUrl || CatItemSvg} defaultSrc={CatItemSvg} width="50" height="50" />
                        </CardHeader>
                        <CardBody className="card_body">
                            <h4>{this.props.name}</h4>
                            {itemDetails(this.props)}
                        </CardBody>
                        <CardFooter>
                        </CardFooter>
                    </div>
                </Card>
            </GridItem>
        );
    };
}

PortfolioItem.propTypes = {
    history: propTypes.object,
    catalog_id: propTypes.string
};

export default withRouter(
    connect(
        null,
        mapDispatchToProps)(PortfolioItem)
);

