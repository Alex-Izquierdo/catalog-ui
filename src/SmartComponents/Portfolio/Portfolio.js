import React, { Component, Fragment } from 'react';
import { withRouter, Route, Switch, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import ContentGallery from '../../SmartComponents/ContentGallery/ContentGallery';
import { fetchSelectedPortfolio, fetchPortfolioItemsWithPortfolio } from '../../redux/Actions/PortfolioActions';
import MainModal from '../Common/MainModal';
import { hideModal, showModal } from '../../redux/Actions/MainModalActions';
import AddProductsToPortfolio from '../../SmartComponents/Portfolio/AddProductsToPortfolio';
import PortfolioFilterToolbar from '../../PresentationalComponents/Portfolio/PortfolioFilterToolbar';
import PortfolioActionToolbar from '../../PresentationalComponents/Portfolio/PortfolioActionToolbar';
import PortfolioItem from './PortfolioItem';
import NoMatch from '../../PresentationalComponents/Shared/404Route';
import AddPortfolioModal from './add-portfolio-modal';
import RemovePortfolioModal from './remove-portfolio-modal';
import { scrollToTop } from '../../Helpers/Shared/helpers';
import './portfolio.scss';

class Portfolio extends Component {
    state = {
      portfolioId: '',
      isKebabOpen: false,
      isOpen: false,
      filteredItems: []
    };

  fetchData = (apiProps) => {
    this.props.fetchSelectedPortfolio(apiProps);
    this.props.fetchPortfolioItemsWithPortfolio(apiProps);
  }

  componentDidMount() {
    this.fetchData(this.props.match.params.id);
    scrollToTop();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.fetchData(this.props.match.params.id);
      scrollToTop();
    }
  }

  onClickEditPortfolio = () => {
    this.props.showModal({
      open: true,
      itemdata: this.props,
      closeModal: this.props.hideModal
    }, 'editportfolio');

    this.setState({
      ...this.state,
      isOpen: !this.state.isOpen
    });
  };

  onClickRemovePortfolio = () => {
    this.props.showModal({
      open: true,
      portfolio: this.props.portfolio,
      onSuccess: this.onPortfolioRemoved,
      closeModal: this.props.hideModal
    }, 'removeportfolio');

    this.setState({
      ...this.state,
      isOpen: !this.state.isOpen
    });
  };

  filterItems = (filterValue) => {
    let filteredItems = [];
    if (this.props.portfolioItems && this.props.portfolioItems.portfolioItems) {
      filteredItems = this.props.portfolioItems.portfolioItems;
      filteredItems = filteredItems.filter(({ name }) => name.toLowerCase().includes(filterValue.toLowerCase()));
    }

    return filteredItems;
  };

  renderProducts = ({ title, filteredItems, addProductsRoute, editPortfolioRoute, removePortfolioRoute }) => (
    <Fragment>
      <PortfolioFilterToolbar/>
      { (!this.props.isLoading) &&
        <div style={ { marginTop: '15px', marginLeft: '25px', marginRight: '25px' } }>
          <PortfolioActionToolbar title={ title }
            onClickEditPortfolio={ this.onClickEditPortfolio }
            onClickRemovePortfolio={ this.onClickRemovePortfolio }
            filterItems={ this.filterItems }
            addProductsRoute={ addProductsRoute }
            editPortfolioRoute={ editPortfolioRoute }
            removePortfolioRoute={ removePortfolioRoute }
          />
        </div>
      }
      <Route exact path="/portfolio/:id/edit-portfolio" component={ AddPortfolioModal } />
      <Route exact path="/portfolio/:id/remove-portfolio" component={ RemovePortfolioModal } />
      <ContentGallery { ...filteredItems } />
      <MainModal/>
    </Fragment>
  )

  renderAddProducts = ({ portfolioRoute }) => (
    <AddProductsToPortfolio
      portfolio={ this.props.portfolio }
      portfolioRoute={ portfolioRoute }
    />
  );

  render() {
    const portfolioRoute = this.props.match.url;
    const addProductsRoute = `${this.props.match.url}/add-products`;
    const editPortfolioRoute = `${this.props.match.url}/edit-portfolio`;
    const removePortfolioRoute = `${this.props.match.url}/remove-portfolio`;
    let filteredItems = {
      items: this.props.portfolioItems.map(item => <PortfolioItem key={ item.id } { ...item }/>),
      isLoading: this.props.isLoading
    };

    let title = this.props.portfolio ? this.props.portfolio.name : '';

    if (this.state.portfolioRemoved) {
      return <Redirect to="/portfolios" />;
    }

    return (
      <Switch>
        <Route path="/portfolio/:id/add-products" render={ props => this.renderAddProducts({ portfolioRoute, ...props }) } />
        <Route
          path="/portfolio/:id"
          render={ props => this.renderProducts({ addProductsRoute, editPortfolioRoute, removePortfolioRoute, filteredItems, title, ...props }) }
        />
        <Route component={ NoMatch } />
      </Switch>
    );
  }
}

const mapStateToProps = ({ portfolioReducer: { selectedPortfolio, portfolioItems, isLoading }}) => ({
  portfolio: selectedPortfolio,
  portfolioItems,
  isLoading: !selectedPortfolio || isLoading
});

const mapDispatchToProps = dispatch => ({
  fetchPortfolioItemsWithPortfolio: apiProps => dispatch(fetchPortfolioItemsWithPortfolio(apiProps)),
  fetchSelectedPortfolio: apiProps => dispatch(fetchSelectedPortfolio(apiProps)),
  hideModal: () => dispatch(hideModal()),
  showModal: (modalProps, modalType) => {
    dispatch(showModal({ modalProps, modalType }));
  }
});

Portfolio.propTypes = {
  isLoading: propTypes.bool,
  fetchPortfolioItemsWithPortfolio: propTypes.func,
  fetchSelectedPortfolio: propTypes.func,
  showModal: propTypes.func,
  hideModal: propTypes.func,
  onClickEditPortfolio: propTypes.func.isRequired,
  onClickRemovePortfolio: propTypes.func.isRequired,
  match: propTypes.object,
  portfolio: propTypes.shape({
    name: propTypes.string,
    id: propTypes.string.isRequired
  }),
  portfolioItems: propTypes.array
};

Portfolio.defaultProps = {
  portfolioItems: []
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Portfolio));
