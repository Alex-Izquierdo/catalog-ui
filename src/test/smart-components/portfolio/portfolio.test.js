import React from 'react';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import configureStore from 'redux-mock-store' ;
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import ToolbarRenderer from '../../../toolbar/toolbar-renderer';
import OrderModal from '../../../smart-components/common/order-modal';
import Portfolio from '../../../smart-components/portfolio/portfolio';
import PortfolioItem from '../../../smart-components/portfolio/portfolio-item';
import { CATALOG_API_BASE, SOURCES_API_BASE } from '../../../utilities/constants';
import FilterToolbarItem from '../../../presentational-components/shared/filter-toolbar-item';
import RemovePortfolioModal from '../../../smart-components/portfolio/remove-portfolio-modal';
import AddProductsToPortfolio from '../../../smart-components/portfolio/add-products-to-portfolio';
import { FETCH_PLATFORMS, FETCH_PORTFOLIO, FETCH_PORTFOLIO_ITEMS_WITH_PORTFOLIO } from '../../../redux/action-types';

describe('<Portfolio />', () => {

  let initialProps;
  let initialState;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;

  const ComponentWrapper = ({ store, initialEntries = [ '/foo' ], children }) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ initialEntries }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      id: '123'
    };
    initialState = {
      platformReducer: {
        platformItems: []
      },
      portfolioReducer: {
        selectedPortfolio: {
          id: '123',
          name: 'Foo'
        },
        portfolioItems: { data: [], meta: {
          limit: 50,
          offset: 0
        }},
        portfolios: { data: [{
          id: '123',
          name: 'bar',
          description: 'description',
          modified: 'sometimes'
        }]}
      }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const wrapper = shallow(<Portfolio { ...initialProps } />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should mount and fetch correct data', async (done) => {
    const store = mockStore(initialState);
    const expectedActions = [{
      type: `${FETCH_PLATFORMS}_PENDING`
    }, {
      type: `${FETCH_PORTFOLIO}_PENDING`
    }, {
      type: `${FETCH_PORTFOLIO_ITEMS_WITH_PORTFOLIO}_PENDING`
    }, expect.objectContaining({
      type: `${FETCH_PLATFORMS}_FULFILLED`
    }), expect.objectContaining({
      type: `${FETCH_PORTFOLIO}_FULFILLED`
    }), expect.objectContaining({
      type: `${FETCH_PORTFOLIO_ITEMS_WITH_PORTFOLIO}_FULFILLED`
    }) ];

    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: {
        application_types: [{
          sources: []
        }]
      }
    }}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: {}}));

    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));

    await act(async() => {
      mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
        </ComponentWrapper>
      );
    });
    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedActions);
      done();
    });
  });

  it('should mount and render add products page', async (done) => {
    const store = mockStore({ ...initialState, platformReducer: { platforms: [], platformItems: {}}});
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123/add-products' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
        </ComponentWrapper>
      );
    });

    setImmediate(() => {
      expect(wrapper.find(AddProductsToPortfolio)).toHaveLength(1);
      done();
    });
  });

  it('should mount and render remove products page and call remove products', async (done) => {
    expect.assertions(2);
    const store = mockStore({
      ...initialState,
      platformReducer: { platforms: []},
      portfolioReducer: {
        ...initialState.portfolioReducer,
        selectedPortfolio: {
          id: '123',
          name: 'Foo'
        },
        portfolioItems: { data: [{
          id: '123',
          name: 'Foo',
          description: 'desc',
          modified: 'sometimes'
        }], meta: {
          limit: 50,
          offset: 0
        }}
      }
    });

    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.delete(`${CATALOG_API_BASE}/portfolio_items/123`, mockOnce((req, res) => {
      expect(req).toBeTruthy();
      done();
      return res.status(200);
    }));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123/remove-products' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
        </ComponentWrapper>
      );
    });

    setImmediate(async () => {
      wrapper.update();
      expect(wrapper.find(ToolbarRenderer)).toHaveLength(1);
      act(() => {
        wrapper.find(PortfolioItem).props().onSelect('123');
        wrapper.update();
      });
      wrapper.find('button#remove-products-dropdown-toggle').simulate('click');
      await act(async () => {
        wrapper.find('li').last().find('span').simulate('click');
      });
    });
  });

  it('should mount and render remove portfolio modal', (done) => {
    const store = mockStore({
      ...initialState,
      platformReducer: { platforms: []},
      portfolioReducer: {
        ...initialState.portfolioReducer,
        portfolioItems: { data: [{
          id: '123',
          name: 'Foo',
          description: 'desc',
          modified: 'sometimes'
        }],
        meta: {
          limit: 50,
          offset: 0
        }}
      }
    });
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));

    const wrapper = mount(
      <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123/remove-portfolio' ] }>
        <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
      </ComponentWrapper>
    );

    setImmediate(() => {
      expect(wrapper.find(RemovePortfolioModal)).toHaveLength(1);
      done();
    });
  });

  it('should mount and render order item modal', async (done) => {
    const store = mockStore({
      ...initialState,
      platformReducer: { platforms: []},
      orderReducer: {
        isLoading: false
      },
      portfolioReducer: {
        ...initialState.portfolioReducer,
        portfolioItems: { data: [{
          id: '123',
          name: 'Foo',
          description: 'desc',
          modified: 'sometimes'
        }],
        meta: {
          limit: 50,
          offset: 0
        }}
      }
    });
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: {
        application_types: [{ sources: []}]
      }
    }}));
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123/order/321' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
        </ComponentWrapper>
      );
    });

    setImmediate(() => {
      expect(wrapper.find(OrderModal)).toHaveLength(1);
      done();
    });
  });

  it('should mount and filter portfolio items', async (done) => {
    const store = mockStore({
      ...initialState,
      platformReducer: { platforms: []},
      portfolioReducer: {
        ...initialState.portfolioReducer,
        portfolioItems: { data: [{
          id: '123',
          name: 'Foo',
          description: 'desc',
          modified: 'sometimes'
        }],
        meta: {
          offset: 50,
          limit: 0
        }}
      }
    });
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/123`, mockOnce({ body: { data: []}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/123' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } { ...args } /> } />
        </ComponentWrapper>
      );
    });

    setImmediate(() => {
      wrapper.update();
      expect(wrapper.find(PortfolioItem)).toHaveLength(1);
      const filterInput = wrapper.find(FilterToolbarItem).first();
      act(() => {
        filterInput.props().onFilterChange('nothing');
      });
      wrapper.update();
      expect(wrapper.find(PortfolioItem)).toHaveLength(0);
      done();
    });
  });

  it('should remove portfolio items and call undo action', async (done) => {
    expect.assertions(1);
    let store = mockStore({
      ...initialState,
      platformReducer: { platforms: []},
      portfolioReducer: {
        ...initialState.portfolioReducer,
        selectedPortfolio: {
          id: '321',
          name: 'Foo'
        },
        portfolioItems: { data: [{
          id: '321',
          name: 'Foo',
          description: 'desc',
          modified: 'sometimes'
        }],
        meta: {
          limit: 50,
          offset: 0
        }}
      }
    });
    const restoreKey = 'restore-321';

    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321`, mockOnce({ body: { data: []}}));
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: []}]}}}));

    /**
     * remove portfolio items calls
     */
    apiClientMock.delete(`${CATALOG_API_BASE}/portfolio_items/321`, mockOnce((_req, res) => res.status(200).body({ restore_key: restoreKey })));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321/portfolio_items?limit=50&offset=0`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321/portfolio_items`, mockOnce({ body: { data: []}}));

    /**
     * undo endpoint
     */
    apiClientMock.post(`${CATALOG_API_BASE}/portfolio_items/321/undelete`, mockOnce((req, res) => {
      expect(JSON.parse(req.body())).toEqual({ restore_key: 'restore-321' });
      done();
      return res.status(200).body({ id: '321' });
    }));

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store } initialEntries={ [ '/portfolios/detail/321' ] }>
          <Route path="/portfolios/detail/:id" render={ (...args) => <Portfolio { ...initialProps } id="321" { ...args } /> } />
        </ComponentWrapper>
      );
    });

    setImmediate(async () => {
      wrapper.update();
      const checkbox = wrapper.find(PortfolioItem).find('input');
      checkbox.simulate('change');
      /**
       * Trigger remove portfolio items action
       */
      const removeTrigger = wrapper.find('button#remove-products-dropdown-toggle');

      /**
       * open dropdown
       */
      removeTrigger.simulate('click');
      wrapper.update();

      /**
       * trigger remove actions
       */
      await act(async() => {
        wrapper.find('li').last().find('span').simulate('click');
      });
      setImmediate(() => {
        /**
         * trigger notification undo click
         */
        const notification = store.getActions()[9].payload.description;
        const notificationWrapper = mount(<IntlProvider locale="en"><React.Fragment>{ notification }</React.Fragment></IntlProvider>);
        notificationWrapper.find('a span').simulate('click');
      });
    });
  });
});
