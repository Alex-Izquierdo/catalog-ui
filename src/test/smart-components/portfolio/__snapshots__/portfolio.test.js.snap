// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portfolio /> should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
